postgres/postgres.go

```diff
@@ -0,0 +1,231 @@
+	addInstance := prepare("INSERT INTO Instances(jobId, instance) VALUES($1, $2) RETURNING instanceId", db)
+	findInstance := prepare("SELECT instance FROM Instances WHERE instanceId = $1", db)
+	updateInstance := prepare("UPDATE Instances set instance = instance || jsonb($1::TEXT) WHERE instanceId = $2 RETURNING instanceId", db)
+	addEvent := prepare("UPDATE Instances SET instance = jsonb_set(instance, '{events}', (SELECT (instance->'events')  || TO_JSONB(json_build_object('type', $1::TEXT, 'time', $2::TEXT, 'message', $3::TEXT, 'messageOffset', $4::TEXT)) FROM Instances WHERE instanceid = $5), true) WHERE instanceid = $5 RETURNING instanceId", db)
+	addDimension := prepare("INSERT INTO Dimensions(instanceId, nodeName, value) VALUES($1, $2, $3)", db)
+	getDimensions := prepare("SELECT nodeName, value, nodeId FROM Dimensions WHERE instanceId = $1", db)
+	addNodeID := prepare("UPDATE Dimensions SET nodeId = $1 WHERE instanceId = $2 AND nodeName = $3 RETURNING instanceId", db)
+	createPublishMessage := prepare("SELECT job->>'recipe', job->'files', STRING_AGG(instanceId::TEXT, ', ') FROM Jobs INNER JOIN  Instances ON (Jobs.jobId = Instances.jobId) WHERE jobs.jobId = $1 GROUP BY jobs.job", db)
+	return Datastore{db: db, addJob: addJob, updateJob: updateJob, addInstance: addInstance, updateInstance: updateInstance,
+		findInstance: findInstance, addFileToJob: addFileToJob, addEvent: addEvent, addDimension: addDimension,
+		getDimensions: getDimensions, addNodeID: addNodeID, createPublishMessage: createPublishMessage}, nil
+}
+
+// AddJob - Add a job to be stored in postgres.
+func (ds Datastore) AddJob(host string, newjob *models.Job) (models.Job, error) {
+	bytes, error := json.Marshal(newjob)
+	if error != nil {
+		return models.Job{}, error
+	}
+	tx, _ := ds.db.Begin()
+	row := tx.Stmt(ds.addJob).QueryRow(bytes)
+	var jobID sql.NullString
+	rowError := row.Scan(&jobID)
+	if rowError != nil {
+		return models.Job{}, rowError
+	}
+
+	id, instanceIDErr := ds.AddInstance(tx, jobID.String)
+	if instanceIDErr != nil {
+		if txError := tx.Rollback(); txError != nil {
+			return models.Job{}, txError
+		}
+		return models.Job{}, instanceIDErr
+	}
+	if txError := tx.Commit(); txError != nil {
+		return models.Job{}, txError
+	}
+	url := host + "/instances/" + id
+	newjob.JobID = jobID.String
+	newjob.Links.InstanceIDs = []string{url}
+	return *newjob, nil
+}
+
+// AddUploadedFile -  Add an uploaded file to a job.
+func (ds Datastore) AddUploadedFile(instanceID string, message *models.UploadedFile) error {
+	row := ds.addFileToJob.QueryRow(message.AliasName, message.URL, instanceID)
+	var returnedInstanceID sql.NullString
+	// Check that a instanceID is returned if not, no rows where update so return a job not found error
+	error := row.Scan(&returnedInstanceID)
+	return convertError(error)
```

Comment:
can once line this `return convertError(row.Scan(&returnedInstanceID))`

This avoids memory allocation